ops/pebble.py

```diff
@@ -1966,3 +1963,177 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+                    break
+                else:
+                    # Insufficient data to continue
+                    break
+
+            if self._part_type == 'response':
+                next_boundary_index = self._get_next_boundary_index()
+                if next_boundary_index is not None:
+                    data = self._buffer[:next_boundary_index]
+                    self._response = json.loads(data.decode())
+                    # Advance the buffer to point at the next boundary
+                    self._buffer = self._buffer[next_boundary_index:]
+                    self._headers = None
+                else:
+                    # Insufficient data to continue
+                    break
+            elif self._part_type == 'files':
+                next_boundary_index = self._get_next_boundary_index()
+                if next_boundary_index is not None:
+                    # Next boundary's location is known; this is the final write.
+                    data = self._buffer[:next_boundary_index]
+                    outfile = self._get_open_tempfile()
+                    outfile.write(data)
+                    outfile.close()
+
+                    # Advance the buffer to point at the next boundary
+                    self._buffer = self._buffer[next_boundary_index:]
+                    self._headers = None
+                else:
+                    # Data present but boundary not yet found.
+                    # Write data from buffer, but don't accidentally write data which may belong
+                    # to the next boundary line.
+                    safe_bound = max(0, len(self._buffer) - len(self._terminal_boundary))
+                    data = self._buffer[:safe_bound]
+
+                    # Don't open/close file for intermediate writes
+                    outfile = self._get_open_tempfile()
+                    outfile.write(data)
+
+                    # Advance the buffer for bytes consumed
+                    self._buffer = self._buffer[safe_bound:]
+
+                    # Insufficient data to continue
+                    break
+
+    def _parse_headers(self, data):
+        parser = email.parser.BytesFeedParser()
+        parser.feed(data)
+        message = parser.close()
+        return message
+
+    def _detect_invalid_content_disposition(self):
+        content_disposition = self._headers.get_content_disposition()
+        if content_disposition != 'form-data':
+            raise RuntimeError(
+                'unexpected content disposition: {}'.format(content_disposition))
+
+    def _prepare_for_part_body(self):
+        name = self._headers.get_param('name', header='content-disposition')
+        if name == 'response':
+            part_type = 'response'
+        elif name == 'files':
+            filename = self._headers.get_filename()
+            if filename is None:
+                raise NotImplementedError('multipart "files" part missing filename')
```

Comment:
That doesn't look like a NotImplementedError either, I think?  Isn't this typically used for Python methods/etc that are missing/abstract?